### DISCUSSION TITLE
MS Help Viewer file support

### DISCUSSION THREAD

**Discussion by msedi** [doxygen/doxygen#10930] [Q&A]
Hello,

are there any plan to support MSHA files for the MS Help File Viewer? I have googled a few sites but couldn't find a request for that?

Thanks

**Comment by paule32** [comment]
they are already support by doxygen 1.10 provided with **GENERATE_HTML** tag.

**Reply by msedi** [reply]
@paule32: Thanks, I saw and tried, but isn't that only to generate chm files? From the docs:

> GENERATE_HTMLHELP
> If the GENERATE_HTMLHELP tag is set to YES then doxygen generates three additional HTML index files: index.hhp, index.hhc, and index.hhk. The index.hhp is a project file that can be read by Microsoft's HTML Help Workshop on Windows. In the beginning of 2021 Microsoft took the original page, with a.o. the download links, offline the HTML help workshop was already many years in maintenance mode). You can download the HTML help workshop from the web archives at [Installation executable](http://web.archive.org/web/20160201063255/http://download.microsoft.com/download/0/A/9/0A939EF6-E31C-430F-A3DF-DFAE7960D564/htmlhelp.exe).
> The HTML Help Workshop contains a compiler that can convert all HTML output generated by doxygen into a single compiled HTML file (.chm). Compiled HTML files are now used as the Windows 98 help format, and will replace the old Windows help format (.hlp) on all Windows platforms in the future. Compressed HTML files also contain an index, a table of contents, and you can search for words in the documentation. The HTML workshop also contains a viewer for compressed HTML files.


But I was looking for the MS Helper Viewer

![image](https://github.com/doxygen/doxygen/assets/8020659/a4cf3edf-c83a-44b1-9890-9f38610c7790)

**Reply by paule32** [reply]
The essential Viewer is out of date a little bit, because Microsoft use the older ActiveX Technology for a HTMLViewer ActiveX.
You can import this ActiveX Type Library (ATL) into your Projects.
If you don't want to use this Technology (because they "speak to home"), and you can use a newer Version of the ActiveX.
Therefore, you have to download the Edge Extension for CHM - it is an update for the old HtmlViewer ActiveX.

If you would like to programming your own Viewer, then take a look to the Gecko Mozilla or Chrome (or Edge) Framework ...
Or, simply implement RTF (Rich Text Format) Editor, and set the Enable Property to False, so the user can not modify the HTML
parts directly.

It give alternatives.

**Comment by doxygen** [comment]
Doxygen supports Compressed HTML Help (even though Microsoft has dropped support for quite some time now).

I did find [this resource](https://learn.microsoft.com/en-us/visualstudio/help-viewer/overview?view=vs-2022) about the Help Viewer, but it is more geared towards the user, not really specifying the format, so it seems to be a proprietary file format.

Feel free to add a better resource.

**Reply by paule32** [reply]
If you a an pascal programmer like me, you can take a look to the FPC (Free Pascal Compiler) and/or Lazarus - the IDE like Delphi.
I don't know exactly (either fpc or lazarus), but they definitively exists a example for writing CHM Project file format.
Please provide more information what your goal is, so I can give you more precise information's.

**Reply by albert-github** [reply]
As far as I know the fpc has a `.chm` generator i.e. `chmcmd` but this works in a similar way as the old Microsoft `hhc`.
In this discussion is is about the `Microsoft Help Viewer` where files have the extension `.mshc` and in essential a zip file with some extra meta tags.

As @doxygen noted the description in https://learn.microsoft.com/en-us/visualstudio/help-viewer/overview?view=vs-2022 is geared to the user, so some questions are open:
- where is a description of the meta tags?
- how can we look at the results once a file has been created?

**Reply by paule32** [reply]
take a look to ...
[using chm files](https://www.aerotech.com/wp-content/uploads/2021/03/Using-Compiled-HTML-Help-files.pdf)
[types of html help](http://docwiki.embarcadero.com/RADStudio/Athens/en/Types_of_HTML_Help)
[pack back](https://stackoverflow.com/questions/41103615/how-to-pack-back-a-chm-file-that-was-unpacked-by-7zip)
[create with doxygen](https://stackoverflow.com/questions/30629770/how-to-generate-chm-with-doxygen-and-html-help-compiler)
[informations](https://learn.microsoft.com/en-us/previous-versions/windows/desktop/htmlhelp/microsoft-html-help-1-4-sdk)
[specification](https://www.nongnu.org/chmspec/latest/index.html)

**Reply by albert-github** [reply]
Thanks for the links though the majority is geared towards CHM which is supported by doxygen and the topic of this discussion, in my opinion, is about `.mshc` so "Microsoft Help 3".

- [using chm files](https://www.aerotech.com/wp-content/uploads/2021/03/Using-Compiled-HTML-Help-files.pdf)
  About CHM
- [types of html help](http://docwiki.embarcadero.com/RADStudio/Athens/en/Types_of_HTML_Help)
  Tells what "Microsoft Help 3" is but no detailed technical information
- [pack back](https://stackoverflow.com/questions/41103615/how-to-pack-back-a-chm-file-that-was-unpacked-by-7zip)
  About CHM
- [create with doxygen](https://stackoverflow.com/questions/30629770/how-to-generate-chm-with-doxygen-and-html-help-compiler)
  About CHM
- [informations](https://learn.microsoft.com/en-us/previous-versions/windows/desktop/htmlhelp/microsoft-html-help-1-4-sdk)
  About HTML Help 1.4 
- [specification](https://www.nongnu.org/chmspec/latest/index.html)
  About CHM

Unfortunately I could not find info about the meta tags and how to view the created files.

**Reply by paule32** [reply]
CHM is the extension of compiled, and compressed HTML file format.
To view the chm file in binary view, I use the Norton Commander like "FAR Manager" for the Microsoft Windows Console.
Before you download, and install, you could take a look to: https://de.wikipedia.org/wiki/FAR_Manager

You start a console, call a script (where your commands are to call doxygen to create needed html files, and to patch this files before you call the hhc.exe - the Microsoft Help compiler).
This can be done within far or plain raw console.

To view the created binary, you start FAR.exe select your project.chm file, and press F3-key.
This enables you to view the binary.

To view the content of a page (that come out of the project.chm), you can open the Explorer.exe by holding the Windows-Key, and pressing the E-Key on your keyboard. Navigate to the path where your chm exists, and double click on this file.
This will automatically open the chm viewer.

Then navigate to a topic, and press the right mouse button.
Then you could use the menu item "Show Site Code" (attention: Show Site Code is my translation to English for showing the site code - also the plain de-compressed HTML file.

This can you do for every topic page/site where you have access within the chm file.
Maybe, you can see your Meta-Data what you wanted for.

My secret Tip:
Open your C-Sharp Application, call the chm viewer, and perform a "SetParent" win32api call, to bring the viewer into your application.
<hr>
For my CHM-Projects, I use **doxygen + helpndoc**.

[HelpNDoc](https://www.helpndoc.com) is an other help authoring Tool that is commercial, but you can a time unlimited personal edition for free.
You can patch doxygen files, to fit your needs, and re-compile it with helpndoc.
For this, you can use a subset of the Pascal programming language for scripting.
As this, I have the following Meta codes:

```
println('[WINDOWS]');
printfln(
  'Main="%s","toc.hhc","%s","%s","%s","%s","%s","%s","%s",%s,%d,%s,[%d,%d,%d,%d],0xB0000,,,%d,0,0,0',
  [
	sProjectTitle,
	sKeywordsFile,
	sDefaultTopicUrl,
	HndBuildsMeta.GetItemMetaStringValue(HndGeneratorInfo.CurrentBuildId, 'WinHomeUrl', sDefaultTopicUrl),
	HndBuildsMeta.GetItemMetaStringValue(HndGeneratorInfo.CurrentBuildId, 'WinJump1Url', ''),
	HndBuildsMeta.GetItemMetaStringValue(HndGeneratorInfo.CurrentBuildId, 'WinJump1Caption', ''),
	HndBuildsMeta.GetItemMetaStringValue(HndGeneratorInfo.CurrentBuildId, 'WinJump2Url', ''),
	HndBuildsMeta.GetItemMetaStringValue(HndGeneratorInfo.CurrentBuildId, 'WinJump2Caption', ''),
	HndBuildsMetaEx.GetChmNavigationPaneStyleHex(HndGeneratorInfo.CurrentBuildId),
	HndBuildsMeta.GetItemMetaIntValue(HndGeneratorInfo.CurrentBuildId, 'WinTabNavWidth', 200),
	HndBuildsMetaEx.GetChmButtonVisibilityHex(HndGeneratorInfo.CurrentBuildId),
	HndBuildsMeta.GetItemMetaIntValue(HndGeneratorInfo.CurrentBuildId, 'WinPosLeft', -1),
	HndBuildsMeta.GetItemMetaIntValue(HndGeneratorInfo.CurrentBuildId, 'WinPosTop', -1),
	HndBuildsMeta.GetItemMetaIntValue(HndGeneratorInfo.CurrentBuildId, 'WinPosRight', -1),
	HndBuildsMeta.GetItemMetaIntValue(HndGeneratorInfo.CurrentBuildId, 'WinPosBottom', -1),
	Integer(not HndBuildsMeta.GetItemMetaBoolValue(HndGeneratorInfo.CurrentBuildId, 'WinTabNavVisible', True))
  ]
);
```
maybe, you can use this informations - what would be great.

**Reply by msedi** [reply]
@albert-github: Thanks for poínting this out. It is indeed about the `MSHC` (also combined with MSHA)  help files and not about `CHM`.
MSHA is the help viewer deliver with Visual Studio, and honestly I don't know if this one can be downloaded at all independently. From what I can see in the link from @doxygen it does not seem to be possible to do an installation without using the visual studio installer.

It seems to be a niche for Visual Studio (maybe a bit of SQL server).

**Reply by paule32** [reply]
the alternative could be:  [Kchmviewer](https://sourceforge.net/projects/kchmviewer/).
It is free and open source, and portable - because it use the Qt Framework (which is available for MacOS, Windows, and Linux).

**Reply by msedi** [reply]
@paule32: Thanks. I think currently I'm suffering the problem that the size of the chm and the index inside is too big so that the file cannot even be created ;-)

**Reply by paule32** [reply]
could be, that the mshc format only support 32-Bit - 2 ^32 - 1 Bytes => 4 GB.
I don't believe that you have such a help file.

But, you could split the data content, and graphical data (if you have many many big pictures in your chm).
One feature is, you can locally hold static text into chm file, and/but refer to external resources like pictures, and/or video medias from a web server.

This makes the size smaller, but don't solve your problem really.

**Reply by albert-github** [reply]
@msedi 
- How do you create the chm file with hhc or chmcmd?
- When not with chmcmd maybe give it a try, maybe it can handle larger files. (I have just the chmcmd.exe on my system that I took from fpc-3.3.1.i386-win32.zip)
- I run the windows l compiler with nmake and cl so I might not have installed everything. Can you see which executable is used to show the content as shown in https://github.com/doxygen/doxygen/discussions/10930#discussioncomment-9718156 (might be an integrated part of an other program that is just using a dll though and not a real standalone program).

**Reply by msedi** [reply]
@paule32: As far as I have read, it is not a problem of the file size. It seems more to be a problem of the large index.

**Reply by msedi** [reply]
@albert-github: I have used hhc. I will try give it a try with chmcmd, but this will take some more time.
The location of the MS Helper Viewer is here: C:\Program Files (x86)\Microsoft Help Viewer\v2.3

I will dig a bit to see if there is also a standalone file/installer.

**Reply by paule32** [reply]
@msedi: I never faced with such a problem. So, I could stuck into cofe brew...
So far as I know, doxygen package comes with a HelpCompiler ?
I have pack the hhc folder and upload it for download [here](https://kallup.net/doxygen/packed.zip) - I do not lay my hand into the fire if it 100% works - because it is a Windows Application (maybe with registry key) - this version is de-packed, no installer !

To create a chm file, you have to set doxygen settings so, that GENERATE_HTMLHELP is set to YES - other switches could be set.
Then, you go to console, navigate to your Doxyfile project, run doxygen, change to the output directory and search for a file called like index.hhp. if it present, then run "hhc.exe index.hhp".

I upload a demo project, that you can use to view how I create doxygen generated chm file.

Be warned: I use a developer Workstation, where different tools are available.
This means, that you have to change the build.bat (a ms-dos console batch file) to your needs.
Because I mix mingw tools with command tools.

I include a filter.py - which is a Python 3 script that I use to filter out/change some parts of the doxygen generated output.
And I use dot.exe - a tool to generate diagrams.

The [demo](https://kallup.net/doxygen/demo.zip) exists in the same folder.

I hope this can help you to create a chm project.

**Reply by albert-github** [reply]
@msedi I found on my system:
- C:\Program Files (x86)\Microsoft Help Viewer\v2.1
- C:\Program Files (x86)\Microsoft Help Viewer\v2.2

and here I see the executables:
- HlpCtntMgr.exe
- HlpViewer.exe

And the `HlpViewer.exe` I can start like `HlpViewer.exe /catalogName VisualStudio14` but it doesn't have content so some more studying (and maybe installing newer versions is required which with MS is always quite difficult...).

**Reply by paule32** [reply]
for me:
```
Application.HelpFile := ExtractFilePath(Application.ExeName) + 'YourCHMFile.chm';
HtmlHelp(0, Application.HelpFile, HH_DISPLAY_TOC, 0);
HtmlHelp(0, Application.HelpFile, HH_DISPLAY_INDEX, DWORD(PWideChar('Some keyword here')));
```
[for Delphi](https://www.drexplain.com/press/articles/how_to_create_help_file_for_delphi_application_in_dr_explain_/)

**Reply by paule32** [reply]
Hi,
You must know, that the CHM and the technology behind comes from 32-Bit area of Windows Computers, including the earlier NT Versions of **Windows 95** (from 1995 - also 30 years backwards).

Before **CHM** there was the **HLP** Format which you could generate through **RTF** (Rich Text Format) styled documents.
This **HLP** Format comes from **Windows 3.0** and **3.1** (also from the 16-Bit area of DOS/Windows Computers.
If you really would like to use **CHM** in your 32-Bit Application's, you can use the Web View ActiveX Component and place it on a Form when using **Delphi 7** 32-Bit or downwards.

But be warned, if you use the free downloadable Versions of Delphi from the open Internet page.
They contain a Trojan virus called **Banker Tr.** - that the most Free Anti-Virus Application can not be detected.
So, you're warned !

It is recommended, you use the RAD Studio Delphi Community Version - they are Free and legal to use.

Thanks for reading
paule32

**Reply by msedi** [reply]
@paule32: I think the discussion is a bit going into the wrong direction. The question was not about using and creatingCHM, it was about 

> support MSHA files for the MS Help File Viewer

Since CHM is so much outdated that you cannot even find methods to create it I would completely leave this out and rather go for something more modern. In the Visual Studio environment it is the MS Help Viewer which is not the CHM Viewer. It uses internally maybe some same technology, But the advantage is that it provides a fast index and is able to search for items.

Creating things with doxygen produced so many files for our project that it was (sorry for that ;-) useless for our purpose, since we are also handing out the help files for our customers.

In our opinion the MS Help Viewer is the best solution for our project/purpose.

Thanks
msedi

**Reply by paule32** [reply]
Okay, I don't often work with Visual Studio C++ Community.
So, the stupid Answers from me.
I apologize me.
One of the reasons that CHM Workshop is not used anymore are the security holes...

**Reply by paule32** [reply]
take a look: [there](https://github.com/doxygen/doxygen/issues/11248#issuecomment-2512070713)
it will contain a suggestion to Edge-WebView2

**Reply by MaxEtMoritz** [reply]
Regarding the MS Help Viewer file format, i could find the following unofficial documentation (archived):
https://web.archive.org/web/20191111192316/http://hv2.helpmvp.com/doc/file-format

for help viewer 1 format, the page above often links to the following page which does no longer exist and is not archived: http://wiki.helpwaregroup.com/ms-help-viewer/hvinfo
But the contents of this page can be accessed under the following URL: https://sites.google.com/helpwaregroup.com/helpware-wiki-kb/ms-help-viewer/hvinfo

so it seems to basically be XHTML 1.0 pages with custom metadata tags, roughly summarized.

**Reply by MaxEtMoritz** [reply]
After some further digging, i actually found an official format description from Microsoft:
https://learn.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2017/extensibility/internals/microsoft-help-viewer-sdk
